src/cargo/ops/cargo_compile.rs

```diff
@@ -413,9 +419,21 @@ impl<'a> CompileFilter<'a> {
         }
     }
 
+    pub fn need_dev_deps(&self) -> bool {
+        match *self {
+            CompileFilter::Default { .. } => false,
+            CompileFilter::Only { examples, tests, benches, .. } =>
+                examples.is_specific() || tests.is_specific() || benches.is_specific()
+        }
+    }
+
     pub fn matches(&self, target: &Target) -> bool {
         match *self {
-            CompileFilter::Default { .. } => true,
+            CompileFilter::Default { .. } => match *target.kind() {
+                TargetKind::Bin => true,
+                TargetKind::Lib(..) => true,
+                _ => false,
+            },
```

Comment:
How come this needed changing?

Reply:
It's not necessary and could be dropped from this PR. But without it, the current situation means I have to give extra flags to `cargo build`, it contradicts the documentation, and it also means the `all-targets` flag is redundant.

Reply:
Hm sorry I missed this earlier, but I'm not sure I quite understand this. Could this perhaps be left out for a future PR?

Reply:
(or maybe a test could be added?)

Reply:
This actually opened up a whole new can of worms, so I've reverted it from this PR and opened up a new issue #5134 which I'll file a PR for as soon as this PR is merged (I have a fix, but it builds on top of this PR).